File: tool/unifmu/resources/backends/python/adder.py
import pickle
from typing import Tuple

from fmi2 import Fmi2FMU
import sys
from fmi2 import Fmi2FMU, Fmi2Status

class Adder(Fmi2FMU):
    def __init__(self,reference_to_attr=None) -> None:
        super().__init__(reference_to_attr)
        self.real_a = 0.0
        self.real_b = 0.0

        self.integer_a = 0
        self.integer_b = 0
        
        self.boolean_a = False
        self.boolean_b = False

        self.string_a = ""
        self.string_b = ""

        self._update_outputs()

        

    def __repr__(self):
        return "Adder"

    def serialize(self):

        bytes = pickle.dumps(
            (
                self.real_a,
                self.real_b,
                self.integer_a,
                self.integer_b,
                self.boolean_a,
                self.boolean_b,
                self.string_a,
                self.string_b,
            )
        )
        return Fmi2Status.ok, bytes

    def deserialize(self, bytes) -> int:
        (
            real_a,
            real_b,
            integer_a,
            integer_b,
            boolean_a,
            boolean_b,
            string_a,
            string_b,
        ) = pickle.loads(bytes)
        self.real_a = real_a
        self.real_b = real_b
        self.integer_a = integer_a
        self.integer_b = integer_b
        self.boolean_a = boolean_a
        self.boolean_b = boolean_b
        self.string_a = string_a
        self.string_b = string_b

        return Fmi2Status.ok

    def _update_outputs(self):
        self.real_c = self.real_a + self.real_b
        self.integer_c = self.integer_a + self.integer_b
        self.boolean_c = self.boolean_a and self.boolean_b
        self.string_c = self.string_a + self.string_b

    def do_step(self, current_time, step_size, no_step_prior):
        
        self._update_outputs()

        return Fmi2Status.ok
        

File: tool/unifmu/resources/backends/csharp/adder.cs
using System.Collections.Generic;
using System;
using System.Globalization;
public class Adder : Fmi2FMU
{
    // Make all class variables properties, in order to access them all in a similar manner. 
    public double real_a { get; set; }
    public double real_b { get; set; }
    public double real_c { get; set; }
    public int integer_a { get; set; }
    public int integer_b { get; set; }
    public int integer_c { get; set; }
    public bool boolean_a { get; set; }
    public bool boolean_b { get; set; }
    public bool boolean_c { get; set; }
    public string string_a { get; set; }
    public string string_b { get; set; }
    public string string_c { get; set; }

    public Adder(Dictionary<uint, string> referenceToAttr) : base(referenceToAttr)
    {
        this.real_a = 0.0f;
        this.real_b = 0.0f;

        this.integer_a = 0;
        this.integer_b = 0;

        this.boolean_a = false;
        this.boolean_b = false;

        this.string_a = "";
        this.string_b = "";

        UpdateOutputs();
    }

    // TODO: implement correctly
    public override (byte[], Fmi2Status) Serialize()
    {
        return (null, Fmi2Status.Ok);
    }

    // TODO: implement correctly
    public override Fmi2Status Deserialize(byte[] state)
    {
        return base.Deserialize(state);
    }

    // Implementation of properties

    // public double real_c
    // {
    //     get { return this.real_a + this.real_b; }
    //     set { }
    // }

    // public int integer_c
    // {
    //     get { return this.integer_a + this.integer_b; }
    //     set { }
    // }

    // public bool boolean_c
    // {
    //     get { return this.boolean_a & this.boolean_b; }
    //     set { }
    // }

    // public string string_c
    // {
    //     get { return this.string_a + this.string_b; }
    //     set { }
    // }

    private void UpdateOutputs()
    {
        this.real_c = real_a + real_b;
        this.integer_c = integer_a + integer_b;
        this.boolean_c = boolean_a && boolean_b;
        this.string_c = string_a + string_b;
    }

    public override Fmi2Status DoStep(double currentTime, double stepSize, bool noStepPrior)
    {
        UpdateOutputs();
        return Fmi2Status.Ok;
    }

}